### ISSUE TITLE
Required parameter functions compatibility with optional parameter functions

### ISSUE THREAD

**Issue by ItaHeld90** [microsoft/TypeScript#32509]
**TypeScript Version:**  3.5.2

**Search Terms:** Optional Parameters, Required Parameters

**Code**

```ts
type Foo = {
	x: number;
};

const condition = false;

// foo is optional
function executeFn(fn: (foo?: Foo) => void) {
	if (condition) {
		fn({ x: 42 });
	} else {
		// calling the handler without foo
		fn();
	}
}

// foo is required
function handleFoo(foo: Foo) {
	// throws an error here, foo wasn't sent (even though it's required)
	console.log(foo.x);
}

// expecting a typescript warning here
executeFn(handleFoo);

```

**Expected behavior:**

Typescript should warn for **providing** a function with a _required parameter_ when a function with an _optional parameter_ is **expected**.

**_Note_**: I understand the behavior for the opposite scenario is correct, i.e: **providing** a function with an _optional parameter_ when a function with a _required parameter_ is **expected**.

**Actual behavior:**

Typescript allows this behavior.
**On my code example:** this behavior leads to a **run-time error** when trying to access a property on the required parameter.

**Playground Link:** 

I turned the _strictNullChecks_ flag off (apparently with the flag on Typescript does warn correctly, though I don't think the two should be related):

https://www.typescriptlang.org/play/?strictNullChecks=false#code/C4TwDgpgBAYg9nKBeKBvAUASAB4C4oB2ArgLYBGEATgNzoC+t6AxnAQM7BQsEAmAlsD6tkUAGYBDADZsIjAPRyxCKHzZQ4YQaynpRRAky0EoEbBCZFgEGAQAUogvnsIA-PnhwAlMgB8UAG5wfDzeGJh8olC23PxGoViYDraoUHhQACwATFB0nrSYdCbS0GGYClxSknwEAOZQwAAW0A3ivJJUUADuAg1wlkpwCUl5WHT06OjlosqqUJQQAI5EfPM8uvqGQsYtbdYIznDuCPFlio2UcJ1qrSaUF5RQTfMANANd4mwEAOScMgSctgg-ggxkafRqDRUwC+anmSxWEBCWG4bDg7QAdJI4DUDujsCMxuhTOZLNY7DseO0PCMgA

**Related Issues:**

**Comment by MartinJohns** [comment]
> I turned the strictNullChecks flag off (apparently with the flag on Typescript does warn correctly, though I don't think the two should be related):

But it is. It changes the check of function type parameters to contravariantly, instead of bivariantly. It's simplified "is `Foo` assignable to `Foo?`" (it is) versus "is `Foo?` assignable to `Foo`" (it is not).

#18654 and https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant

**Comment by ItaHeld90** [comment]
thanks for the explanation and the quick reply.

> It is no different from calling your function and providing no value: handleFoo(undefined)

The behavior that's weird in my example, is that it's possible to practically call handleFoo without sending the undefined **explicitly**.

but maybe it is the strictNullChecks flag that should protect against this potential confusion.

**Comment by RyanCavanaugh** [comment]
Allowing `undefined`/`null` as part of any type is what `strictNullChecks` off mode does, and the example does exactly that.
